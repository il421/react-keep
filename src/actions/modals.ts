import {
  CollapseType,
  Modal,
  ModalActionsTypes,
  SetSidebarCollapseAction,
  ToggleModalAction,
} from "../store/store.types";

export const toggle = (modal: Modal, isOpen: boolean): ToggleModalAction => ({
  type: ModalActionsTypes.toggle,
  modal,
  isOpen,
});

export const setCollapsedOptionsInSidebar = (
  collapsed: CollapseType[]
): SetSidebarCollapseAction => ({
  type: ModalActionsTypes.setCollapsedOptions,
  collapsed,
});
