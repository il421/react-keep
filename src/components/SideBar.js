import React from 'react';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Tags from './tags/Tags';

export const SideBar = ({ hideSidebar, showBar }) => {
  return (
    <>
      <div className={ `sidebar ${showBar && 'sidebar--show'}` }>
        <div className="sidebar__close">
          <button className="button--sidenav" onClick={() => hideSidebar()}>
            <FontAwesomeIcon icon="times" size="3x"/>
          </button>
        </div>

        <Tags hideSidebar={() => hideSidebar()} />
      </div>

      {
        showBar && <div className="sidebar__cover" onClick={ () => hideSidebar() } />
      }
    </>
  );
};

export  default SideBar;
