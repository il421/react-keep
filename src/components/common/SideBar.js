import React from 'react';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Tags from '../tags/Tags';

export const SideBar = ({ hideSidebar, showBar }) => {
  return (
    <>
      <div className={ `sidebar ${showBar && 'sidebar--show'}` }>
        <div className="sidebar__close">
          <button className="button--sidenav sidebar__close-button" onClick={() => hideSidebar()}>
            <FontAwesomeIcon icon="times" size="2x"/>
          </button>
        </div>

        <Tags hideSidebar={() => hideSidebar()} />
      </div>

      {
        showBar && <div className="sidebar__cover" onClick={ () => hideSidebar() } />
      }
    </>
  );
};

export  default SideBar;
