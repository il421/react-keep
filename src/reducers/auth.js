export default (state = {}, action) => {
  switch (action.type) {
  case 'LOGIN':
    return {
      uid: action.uid,
      name: action.name,
      url: action.url
    };
  case 'LOGOUT':
    return {};
  default:
    return state;
  }
};
