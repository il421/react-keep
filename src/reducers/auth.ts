import {
  AuthActionsTypes,
  AuthStoreState,
  LoadingAction,
  LoginAction,
  LogoutAction
} from "../store/store.types";

type AuthAction = LoginAction | LogoutAction | LoadingAction;

export default (
  state = {} as AuthStoreState,
  action: AuthAction
): AuthStoreState | object => {
  switch (action.type) {
    case AuthActionsTypes.login:
      return {
        ...state,
        uid: action.uid,
        name: action.name,
        url: action.url
      } as AuthStoreState;

    case AuthActionsTypes.logout:
      return {};

    case AuthActionsTypes.loading:
      return {
        ...state,
        loading: action.loading
      } as AuthStoreState;

    default:
      return state;
  }
};
