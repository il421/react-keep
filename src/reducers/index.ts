export * from "./auth";
export * from "./collaborators";
export * from "./notes";
export * from "./tags";
export * from "./filters";
