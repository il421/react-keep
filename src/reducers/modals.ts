import {
  ModalActionsTypes,
  ModalsStoreState,
  SetSidebarCollapseAction,
  ToggleModalAction,
} from "../store/store.types";
import { NoteType } from "../components/notes";

export const modalsReducerDefaultState: ModalsStoreState = {
  user: {
    isOpen: false,
  },
  sidebar: {
    isOpen: false,
    collapsed: [],
  },
  note: {
    isOpen: false,
    type: NoteType.text,
  },
};

type ModalsAction = ToggleModalAction | SetSidebarCollapseAction;

export default (
  state: ModalsStoreState = modalsReducerDefaultState,
  action: ModalsAction
) => {
  switch (action.type) {
    case ModalActionsTypes.toggle:
      return {
        ...state,
        [action.modal]: {
          ...state[action.modal],
          isOpen: action.isOpen,
        },
      };

    case ModalActionsTypes.setCollapsedOptions:
      return {
        ...state,
        sidebar: {
          ...state.sidebar,
          collapsed: action.collapsed,
        },
      };

    default:
      return state;
  }
};
