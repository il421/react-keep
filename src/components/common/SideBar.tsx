import React from "react";
import { IconButton, FlexBox } from "../ui-components";
import "../../styles/components/common/_sidebar.scss";
import { JustifyContent, toggleArrayElement } from "../../common";
import Tags from "../tags/Tags";
import { ArchiveList } from "../archive/ArchiveList";
import { Collaborators } from "../collaborators/Collaborators";
import { ThunkDispatch } from "redux-thunk";
import {
  CollapseType,
  Modal as ModalType,
  ModalsStoreState,
  Store,
} from "../../store/store.types";
import { setCollapsedOptionsInSidebar, toggle } from "../../actions/modals";
import { connect } from "react-redux";

interface SideBarProp {}

interface StateProps {
  modals: ModalsStoreState;
}

interface DispatchProps {
  toggle: (modal: ModalType, isOpen: boolean) => void;
  setCollapsedOptionsInSidebar: (collapsed: CollapseType[]) => void;
}

export type SideBarBaseProps = SideBarProp & DispatchProps & StateProps;

export const SideBarBase: React.FunctionComponent<SideBarBaseProps> = ({
  modals,
  toggle,
  setCollapsedOptionsInSidebar,
}) => {
  const handleCloseModal = () => toggle("sidebar", false);
  const handleSetCollapsed = (item: CollapseType) => () =>
    setCollapsedOptionsInSidebar(toggleArrayElement(collapsed, item));

  const { isOpen, collapsed } = modals.sidebar;
  return (
    <>
      <div className={`sidebar ${isOpen && "sidebar--show"}`}>
        <FlexBox justifyContent={JustifyContent.end}>
          <IconButton
            className="sidebar__close-button"
            onButtonClick={handleCloseModal}
            icon="times"
            size="2x"
          />
        </FlexBox>

        <div className="dashboard__sidebar">
          <>
            <IconButton
              icon="tags"
              text="Tags list"
              onButtonClick={handleSetCollapsed(CollapseType.tags)}
              className={
                collapsed.includes(CollapseType.tags)
                  ? "icon-button--collapsed"
                  : undefined
              }
            />
            {collapsed.includes(CollapseType.tags) && <Tags />}
          </>

          <>
            <IconButton
              icon="archive"
              text="Archived notes"
              onButtonClick={handleSetCollapsed(CollapseType.arch)}
              className={
                collapsed.includes(CollapseType.arch)
                  ? "icon-button--collapsed"
                  : undefined
              }
            />
            {collapsed.includes(CollapseType.arch) && <ArchiveList />}
          </>
          <>
            <IconButton
              icon="user-friends"
              text="Collaborators"
              onButtonClick={handleSetCollapsed(CollapseType.collaborators)}
              className={
                collapsed.includes(CollapseType.collaborators)
                  ? "icon-button--collapsed"
                  : undefined
              }
            />
            {collapsed.includes(CollapseType.collaborators) && (
              <Collaborators />
            )}
          </>
        </div>
      </div>

      {
        // handle click on cover to close the sidebar
        isOpen && <div className="sidebar__cover" onClick={handleCloseModal} />
      }
    </>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  toggle: (modal: ModalType, isOpen: boolean) =>
    dispatch(toggle(modal, isOpen)),
  setCollapsedOptionsInSidebar: (collapsed: CollapseType[]) =>
    dispatch(setCollapsedOptionsInSidebar(collapsed)),
});

const mapStateToProps = (state: Store): StateProps => {
  return {
    modals: state.modals,
  };
};

export const SideBar = connect<StateProps, DispatchProps, SideBarProp, Store>(
  mapStateToProps,
  mapDispatchToProps
)(SideBarBase);
