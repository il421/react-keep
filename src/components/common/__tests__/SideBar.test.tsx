import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { SideBarBase, SideBarBaseProps } from "../SideBar";
import { modalsReducerDefaultState } from "../../../reducers/modals";

let props: SideBarBaseProps, wrapper: ShallowWrapper<SideBarBaseProps, any>;

beforeEach(() => {
  props = {
    modals: modalsReducerDefaultState,
    toggle: jest.fn(),
    setCollapsedOptionsInSidebar: jest.fn(),
  };

  wrapper = shallow<SideBarBaseProps>(<SideBarBase {...props} />);
});

test("should render SideBar", () => {
  expect(wrapper.exists(".sidebar__cover")).toBeFalsy();
  expect(wrapper.debug()).toMatchSnapshot();
});

test("should close side bar if click on close button", () => {
  wrapper.find("IconButton").at(0).prop<any>("onButtonClick")();
  expect(props.toggle).toHaveBeenLastCalledWith("sidebar", false);
});

test("should close side bar if click on cover", () => {
  wrapper = shallow<SideBarBaseProps>(
    <SideBarBase
      {...props}
      modals={{
        ...modalsReducerDefaultState,
        sidebar: { collapsed: [], isOpen: true },
      }}
    />
  );
  wrapper.update();
  expect(wrapper.exists(".sidebar__cover")).toBeTruthy();
  wrapper.find(".sidebar__cover").simulate("click");
  expect(props.toggle).toHaveBeenLastCalledWith("sidebar", false);
});
