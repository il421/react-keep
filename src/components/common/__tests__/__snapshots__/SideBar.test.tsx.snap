// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SideBar 1`] = `
"<Fragment>
  <div className=\\"sidebar false\\">
    <FlexBox justifyContent=\\"flex-end\\">
      <IconButton className=\\"sidebar__close-button\\" onButtonClick={[Function: handleCloseModal]} icon=\\"times\\" size=\\"2x\\" />
    </FlexBox>
    <div className=\\"dashboard__sidebar\\">
      <IconButton icon=\\"tags\\" text=\\"Tags list\\" onButtonClick={[Function]} className={[undefined]} />
      <IconButton icon=\\"archive\\" text=\\"Archived notes\\" onButtonClick={[Function]} className={[undefined]} />
      <IconButton icon=\\"user-friends\\" text=\\"Collaborators\\" onButtonClick={[Function]} className={[undefined]} />
    </div>
  </div>
</Fragment>"
`;
